### PR TITLE
feat: non preemptive Global EDF scheduler

### PR DESCRIPTION
**Pull request by kobayu858**
## Description

- Non-preemptive GEDF scheduler.
- Using BTreeSet.
- Nodes do not have individual release offsets and deadlines when scheduled by the GEDF scheduler; they share the same absolute deadline of their jobs.  (paper III Task Model lower left)

## Related links

[EmbIV paper](https://drive.google.com/file/d/12WQBW6H6MM2lGVNGS2ohVS1Q1f3rqxKc/view?usp=drive_link)
[EmbIV summary](https://docs.google.com/presentation/d/1v3qpp-YaFCtC1IoPpE68A0oNZPJ7F690/edit?usp=drive_link&ouid=114908281052440997053&rtpof=true&sd=true)
[OpenLink BTreeMap](https://runebook.dev/ja/docs/rust/std/collections/struct.btreemap).

## Pre-review checklist for the PR author

- [x] I've confirmed the [self-checklist](https://docs.google.com/spreadsheets/d/1p5KMcm752iCySHJz0xL1qf6fcLDKprbo/edit#gid=863618628).
- [x] The PR is ready for review.

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

### CODE REVIEW

File: lib/src/global_edf_scheduler.rs [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+                    let self_dag = self.node_data.params.get(key2).unwrap();
+                    let other_dag = other.node_data.params.get(key2).unwrap();
+                    Some(self_dag.cmp(other_dag))
+                }
+                other => other,
+            },
+            other => Some(other),
+        }
+    }
+}
```
Comment (atsushi421) [inline lib/src/global_edf_scheduler.rs:36]:
```suggestion
impl PartialOrd for NodeDataWrapper {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        let comparison_metric = "absolute_deadline";

        match self
            .node_data
            .get_params_value(comparison_metric)
            .cmp(&other.node_data.get_params_value(comparison_metric))
        {
            // If the keys are equal, compare by id
            Ordering::Equal => match self.node_data.id.partial_cmp(&other.node_data.id) {
                // If the ids are also equal, compare by dag_id
                Some(Ordering::Equal) => Some(
                    self.node_data
                        .get_params_value("dag_id")
                        .cmp(&other.node_data.get_params_value("dag_id")),
                ),
                other => other,
            },
            other => Some(other),
        }
    }
}
```

File: lib/src/global_edf_scheduler.rs [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+impl Ord for NodeDataWrapper {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.partial_cmp(other).unwrap_or(Ordering::Equal)
+    }
+}
```
Comment (atsushi421) [inline lib/src/global_edf_scheduler.rs:42]: This process should be moved to scheduler.rs.

File: lib/src/graph_extension.rs [hunk 3]
```diff
@@ -50,11 +61,12 @@ pub trait GraphExtension {
     fn get_anc_nodes(&self, node_i: NodeIndex) -> Option<Vec<NodeIndex>>;
     fn get_des_nodes(&self, node_i: NodeIndex) -> Option<Vec<NodeIndex>>;
     fn get_parallel_process_nodes(&self, node_i: NodeIndex) -> Option<Vec<NodeIndex>>;
-    fn get_dag_id(&self) -> usize;
-    fn set_dag_id(&mut self, dag_id: usize);
+    fn get_dag_value(&self, key: &str) -> i32;
+    fn set_dag_value(&mut self, key: &str, value: i32);
```
Comment (atsushi421) [inline lib/src/graph_extension.rs:65]:
```suggestion
    fn get_dag_param(&self, key: &str) -> i32;
    fn set_dag_param(&mut self, key: &str, value: i32);
```

**Review by atsushi421 — CHANGES_REQUESTED** [review]

File: lib/src/global_edf_scheduler.rs (hunk 2 above)
Comment (kobayu858) [reply lib/src/global_edf_scheduler.rs:42]: Moved.

File: lib/src/graph_extension.rs (hunk 3 above)
Comment (kobayu858) [reply lib/src/graph_extension.rs:65]: Changed.

File: lib/src/global_edf_scheduler.rs (hunk 1 above)
Comment (kobayu858) [reply lib/src/global_edf_scheduler.rs:36]: Thanks for the correction. Changed.

**Review by atsushi421 — APPROVED** [review]
LGTM

**Pull request merged**